bpo-31993: do not allocate large temporary buffers in pickle dump

ogrisel opened the pull request:
For all protocols: avoid concatenating large bytes and str with their opcode
and size header but instead issue an individual call to self.write(data).

For protocol 4: if the size of the opcode + size header + data is larger than
the target frame size, commit the current frame and bypass io.BytesIO to write
the next frame directly to the underlying file object.


<!-- issue-number: bpo-31993 -->
https://bugs.python.org/issue31993
<!-- /issue-number -->

Edit: this now includes changes both the Python and C implementations of the picklers.

ogrisel committed:
Do no allocate large temp buffers in _Pickler.dump

For all protocols: avoid concatenating large bytes and str with their opcode
and size header but instead issue an individual call to self.write(data).

For protocol 4: if the size of the opcode + size header + data is larger than
the target frame size, commit the current frame and bypass io.BytesIO to write
the next frame directly to the underlying file object.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

pitrou commented on Lib/pickle.py:
```diff
@@ -216,6 +216,21 @@ def write(self, data):
         else:
             return self.file_write(data)
 
+    def write_chunks(self, *chunks):
```
Nit: I would call this `write_many`.

pitrou commented on Lib/pickle.py:
```diff
@@ -216,6 +216,21 @@ def write(self, data):
+        if self.current_frame and total_size >= self._FRAME_SIZE_TARGET:
+            # Terminate the current frame to write the next frame directly into
+            # the underlying file to skip the unnecessary memory allocations
+            # of a large temporary buffers.
```
Typo: "a large temporary buffers".

pitrou commented on Lib/pickle.py:
```diff
@@ -216,6 +216,21 @@ def write(self, data):
+        else:
+            write = self.write
+
+        for chunk in chunks:
```
Add a comment pointing out that you avoid concatenation?

serhiy-storchaka commented on Lib/pickle.py:
```diff
@@ -694,11 +710,11 @@ def save_bytes(self, obj):
         n = len(obj)
         if n <= 0xff:
-            self.write(SHORT_BINBYTES + pack("<B", n) + obj)
+            self._write_many(SHORT_BINBYTES, pack("<B", n), obj)
```
I don't think this is needed for short bytes and strings.

ogrisel committed:
Add comment to justify iteration over chunks

serhiy-storchaka commented on Lib/pickle.py:
```diff
@@ -694,11 +710,11 @@ def save_bytes(self, obj):
+            self._write_many(SHORT_BINBYTES, pack("<B", n), obj)
         elif n > 0xffffffff and self.proto >= 4:
-            self.write(BINBYTES8 + pack("<Q", n) + obj)
+            self._write_many(BINBYTES8, pack("<Q", n), obj)
```
It may be better to concatenate the opcode and the size.

ogrisel committed:
Concatenate short bytes

ogrisel commented:
Thanks for the quick review.

I think I addressed your comments. I am not sure how to write the entry in `NEWS.d`. I think this PR is useful as it is and we should not wait for the C version. Shall I add an entry for the Python _Pickler alone in the NEWS file for 3.7.0a2?

I would appreciate it if someone else could do the C version.

pitrou commented:
Yes, you should add a NEWS entry.  You can (and should) use the ["blurb" utility](https://pypi.python.org/pypi/blurb) for that.

I think it's fine to leave the C version for later or to someone else if you're not comfortable with that.

ogrisel committed:
Add entry to NEWS

ogrisel commented:
Done. Feel free to edit it directly if you want to change anything.

ogrisel committed:
Reduce overhead for small bytes objects

ogrisel committed:
Even more overhead reduction for small bytes buffers

ogrisel committed:
Fix no-copy for protocol < 4

ogrisel committed:
More overhead reduction for small bytes

serhiy-storchaka commented:
I'm not sure about 5d463159bfecd5fafb4745e63c0f24c6d0815fdf. opcode and size_header are concatenated in any case. If there is any benefit from this change, how can it be explained?

serhiy-storchaka commented:
Ah, you already have reverted this change. The final variant looks pretty simple.

ogrisel commented:
Yes sorry I had pushed a bad intermediate version and then rebased on top of master to make sure I was benchmarking against the same master as yours. I think the intermediate commits are useless, do you want me to squash them all into a single commit?

serhiy-storchaka commented:
Don't squash them. It is more convenient to make a review if intermediate commits are not squashed. All them will be squashed when be merged in master.

ogrisel commented:
Alright.

ogrisel committed:
Direct write for large bytes in the C-pickler

ogrisel committed:
Update NEWS to include the C pickler

ogrisel commented:
@serhiy-storchaka approved this PR but I subsequently included an implementation for the C version that has not been reviewed yet. The "awaiting-merge" label should be removed and replaced by "awaiting-review".

pitrou commented on Modules/_pickle.c:
```diff
@@ -2054,6 +2055,43 @@ save_float(PicklerObject *self, PyObject *obj)
 static int
+_Pickler_write_large_bytes(
+    PicklerObject *self, const char *header, Py_ssize_t header_size,
+    PyObject *payload, Py_ssize_t payload_size)
```
Why don't you simply pass the payload as a `const char *` instead?

pitrou replied:
Oh, I see, it's because you pass it to `self->write`.

ogrisel committed:
No-copy dump for large unicode in C pickler

ogrisel committed:
Update NEWS to mention str [ci skip]

ogrisel commented on Modules/_pickle.c:
```diff
@@ -890,14 +890,15 @@ _Pickler_WriteFrameHeader(PicklerObject *self, char *qdata, size_t frame_len)
 
 static int
-_Pickler_CommitFrame(PicklerObject *self)
+_Pickler_CommitFrame(PicklerObject *self, size_t frame_extension)
```
I assumed that this is not part of the public API so it's fine to change the prototype of this function without caring about backward compat. Let me know if this is not the case.

serhiy-storchaka replied:
All static functions are private.

ogrisel commented:
Do you want me to convert [this script](https://gist.github.com/ogrisel/0e7b3282c84ae4a581f3b9ec1d84b45a) into a new non-regression test that ensures that future changes to the pickle module will not break the no-copy semantics dump for large `bytes` & utf-8 backed `str`?

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -2054,6 +2055,43 @@ save_float(PicklerObject *self, PyObject *obj)
+    if (_Pickler_FlushToFile(self) < 0)
+        return -1;
+
+    // Reinitialize the buffer for subsequent calls to _Pickler_Write.
```
Some of your comments ends with a period, but others do not. Would be better to be uniform.

serhiy-storchaka commented on Lib/pickle.py:
```diff
@@ -216,6 +216,22 @@ def write(self, data):
+            # allocations of a large temporary buffer.
+            self.commit_frame(force=True)
+            frame_size = len(header) + len(payload)
+            write(FRAME + pack("<Q", frame_size))
```
Is it worth to create a new frame containing a single bytes/string object? This will increase the size of the file, but will not decrease the number of file reading operations (opcode + frame/bytes/string size + payload).

ogrisel replied:
I think the presence of a frame is required by protocol 4. The overhead is likely to be small because this code path is only there for large enough bytes instances.

ogrisel replied:
[PEP 3154](https://www.python.org/dev/peps/pep-3154/#framing) is not very explicit whether framing is required or optional. The existing implementations (Python and C) would never dump a frameless opcode as far as I understand so it might be safer to follow that behavior in the new nocopy code path but I can change that if you think this is being too conservative.

pitrou replied:
I remember that some people voiced concerns about the frame header overhead (however small it is), so I worded the PEP so that it isn't made mandatory (and the unpickler should be able to handle both with and without frames). But I do think the standard pickler implementations should always emit frames in protocol 4.

serhiy-storchaka replied:
Frames are optional.

If don't create a frame for a separate bytes object, this could decrease memory usage on unpickling. You could directly use the result of the `read()` method without creating a buffer.

ogrisel replied:
Indeed I observed at least one useless memory copy at load time for large objects. I have not investigated yet, however, it's there both with the frameless protocol 3 and with protocol 4 with framing enabled. Anyway, it would be best to choose a design that allows for no-copy semantics of large objects both at dump and load times.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2017-11-10-00-05-08.bpo-31993.-OMNg8.rst:
```diff
@@ -0,0 +1,3 @@
+The C-based and Python-based picklers do no longer allocate temporary memory
```
There are no other picklers except the C-based and Python-based picklers. :smile_cat:

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -2054,6 +2055,43 @@ save_float(PicklerObject *self, PyObject *obj)
+    PyObject *result;
+
+    // Commit the previous frame
+    if (_Pickler_CommitFrame(self, 0))
```
New PEP 7 rule requires braces even around the `if` body containing a simple `return` statement. The code in this module is older than this rule, but the new code should obey it.

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -2054,6 +2055,43 @@ save_float(PicklerObject *self, PyObject *obj)
+    if (_Pickler_CommitFrame(self, 0))
+        return -1;
+
+    // Create a new frame dedicated to the large bytes
```
The same doubt as about the Python code. Is creating a new frame needed here?

serhiy-storchaka reviewed (commented):
Excellent! I didn't expect that you will go so far! And the resulting code is pretty simple.

Added few style comments and one question.

ogrisel committed:
C-code style fixes

ogrisel committed:
Simpler NEWS entry [ci skip]

ogrisel commented:
The CI / github integration seems to be confused by the [ci skip]. I will push again my last 2 commits in reverse order.

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -2293,19 +2303,23 @@ write_utf8(PicklerObject *self, const char *data, Py_ssize_t size)
-        if (_Pickler_write_large_bytes(self, header, len, mem, size) < 0)
+        }
+        if (_Pickler_write_large_bytes(self, header, len, mem, size) < 0) {
             return -1;
```
Leaked `mem`.

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -2063,30 +2065,35 @@ _Pickler_write_large_bytes(
     result = PyObject_CallFunctionObjArgs(self->write, payload, NULL);
     Py_XDECREF(result);
-    if (result == NULL)
+    if (result == NULL) {
```
Comparing a pointer to a freed memory with NULL is an undefined behavior.

You don't need to decref it if it is NULL.

serhiy-storchaka reviewed (changes requested):
There is a leak and UB.

ogrisel committed:
Fixed leak

ogrisel committed:
Do not wrap large objects in frames

ogrisel committed:
Fix bug in function call result handling

ogrisel committed:
Fix frameless blobs test for pickletools.optimize

ogrisel committed:
Flush to file after each frame commit

ogrisel commented:
I have made the requested changes; please review again.

In particular:
- Large objects are no longer wrapped in frames.
- I have implemented the "flush after frame commit" behavior for the C pickler and tried make the Python pickler behave consistently (check the new test: `test_framed_write_sizes`.

As a consequence, it means that the C implementation of protocol 3 makes a single write (at the end of dump) while protocol 4 can issue many calls to write (`total_size / 64 kB` approximately). Maybe we should consider adding an option to control the target frame size in protocol 4. Also, we might want to make it larger by default to avoid introducing any performance regression on typical workloads.

ogrisel committed:
Extend NEWS entry to mention write-on-frame-commit

ogrisel commented:
I rebased because the CI status was stalled again.

pitrou commented on Lib/pickle.py:
```diff
@@ -203,19 +204,45 @@ def commit_frame(self, force=False):
-                    write(data)
+                    po = self.delayed_proto_opcode
+                    if po is None:
+                        self.file_write(FRAME + pack("<Q", n) + data)
```
I don't think you should concatenate `data` to the rest. It's not a huge string but it's still 64 kB or more.

pitrou commented on Lib/pickle.py:
```diff
@@ -203,19 +204,45 @@ def commit_frame(self, force=False):
 
+    def write_proto(self, data):
+        if self.current_frame:
+            # Store proto opcode to delay write when committing the first
```
Can you explain why this is suddenly needed?

ogrisel replied:
I wanted to make the Python version issue the same number of calls to `file.write` as the C version does, that is only 1 call to write per 64kb frame. But this is probably a bad idea. Let me change that back.

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -952,6 +939,33 @@ _Pickler_FlushToFile(PicklerObject *self)
+                return -1;
+            }
+        }
+        return 0;
```
This can be moved outside of the `if` block.

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -890,42 +890,29 @@ _Pickler_WriteFrameHeader(PicklerObject *self, char *qdata, size_t frame_len)
 
 static int
-_Pickler_CommitFrame(PicklerObject *self)
+_Pickler_CommitFrame(PicklerObject *self, size_t frame_extension)
```
Isn't `frame_extension` always 0 now?

pitrou commented on Modules/_pickle.c:
```diff
@@ -952,6 +939,33 @@ _Pickler_FlushToFile(PicklerObject *self)
+        if(_Pickler_CommitFrame(self, 0)) {
+            return -1;
+        }
+        if (self->write != NULL) {
```
Should this block be part of `_Pickler_CommitFrame` instead?

ogrisel replied:
Actually `_Pickler_FlushToFile` calls `_Pickler_GetString` which calls `_Pickler_CommitFrame` (that does nothing if the frame has already been committed).

We always need to `_Pickler_CommitFrame` here, but flushing should only be done if we are writing to a file (which is not always the case, e.g. when calling `dumps`).

I could refactor those three functions but I was afraid to introduce subtle semantic changes.

serhiy-storchaka commented on Lib/test/pickletester.py:
```diff
@@ -2125,6 +2168,37 @@ def remove_frames(pickled, keep_frame=None):
             self.assertEqual(obj, self.loads(some_frames_pickle))
 
+    def test_framed_write_sizes(self):
+        if not hasattr(self, 'pickler'):
```
Could you explain what tests are skipped?

ogrisel replied:
Some test classes do not have a pickler attribute and just uses the `self.dumps` / `self.loads` methods. For those, we cannot change the `fast` flag to disable memoization.

serhiy-storchaka replied:
It seems to me that currently all test classes with the `test_framed_write_sizes` method have the `pickler` attribute.

The danger of skipping the test if not hasattr(self, 'pickler'), is that after changing the name of the `pickler` attribute **all** tests will be silently skipped.

ogrisel replied:
`test.test_pickletools.OptimizedPickleTests` does not have it.

ogrisel replied:
I have removed the skip to stay on the safe side. At the cost of running the test for `fast=False` twice.

ogrisel replied:
Hum this one has not been updated. Let me fix that.

ogrisel replied:
Actually, for this test I am not sure what to do: `OptimizedPickleTests` does not have `self.pickler` but this test only makes sense when the test case exposes a `self.pickler` class. It cannot run with the `self.dumps` attribute only.

serhiy-storchaka replied:
It uses `self.pickler` only in `self.pickler(w, proto).dump(small_objects)`. Why not use just `self.dump(small_objects)` here?

ogrisel replied:
Because the goal of this test is to measure the size of the chunks written into a custom writer object. This cannot be done with the `dumps` API.

serhiy-storchaka replied:
Ah, in this case I would skip the test in `OptimizedPickleTests`. But it is more reliable to set `test_framed_write_sizes = None` in `OptimizedPickleTests` instead of testing `hasattr(self, 'pickler')`. If the `pickler` attribute will be renamed this test will loudly fail instead of became silently skipped.

serhiy-storchaka commented on Lib/test/test_pickletools.py:
```diff
@@ -7,6 +7,10 @@
 
+    # pickletools.optimize only works in-memory on pickle strings: it
+    # therefore is ok to include large objects inside a large frame.
+    frameless_blobs = False
```
I think it would be better to write frameless blobs in `pickletools.optimize()`. This will decrease a peak memory on unpickling.

ogrisel replied:
I will try to do that now.

serhiy-storchaka commented on Lib/test/pickletester.py:
```diff
@@ -2076,15 +2095,39 @@ def test_framing_many_objects(self):
+                    self.assertEqual(obj, unpickled)
+                    n_frames = count_opcode(pickle.FRAME, pickled)
+                    if not fast:
+                        # One frame per memoize for each large object.
```
I suggest to not write a frame for a data <= 3 bytes. Reading a frame requires 3 reading operations, and 3 bytes can be read by at most 3 operations. There is no benefit of using a frame.

serhiy-storchaka replied:
I'm going to create a separate issue for more accurate optimization of frames, but maybe the simpler heuristic (size <= 3) can be used in this PR. If this will not complicate the code.

pitrou replied:
I'd rather see the new heuristics in a separate PR personally :-)

ogrisel replied:
I'd also rather do it in a separate PR as it's not trivial to remove a frame header from the `output_buffer` a posteriori.

ogrisel commented:
Thanks for the review. I am working on other issues today and might be a bit busy on other things in the coming days but I plan to address them as soon as possible.

ogrisel committed:
Make proto 4 Python pickler issue (2 * n_frames + 1) calls to write

ogrisel committed:
Cleanup _Pickler_CommitFrame

ogrisel committed:
Get rid of the frameless_blobs=False exception: actually it works out of the box

ogrisel committed:
Implement frameless blobs for pickletools.optimize

ogrisel committed:
Do not skip test silently if self.pickler is renamed

ogrisel committed:
Remove on attribute lookup

ogrisel commented:
I think I addressed all the comments. Do you want me to add a memory test to protect against any future regression in the new no-copy semantics of the `dump` methods?

pitrou commented:
> Do you want me to add a memory test to protect against any future regression in the new no-copy semantics of the dump methods?

This is just an optimization, not a semantic guarantee. Besides, it's quite tricky to measure memory consumption in a reliable and cross-platform way.

serhiy-storchaka commented on Lib/pickletools.py:
```diff
@@ -2251,7 +2251,7 @@ def genops(pickle):
 # A pickle optimizer.
 
-def optimize(p):
+def optimize(pickled):
```
Why not `pickle`? In any case I would left renaming the argument to other issue. There are many related changes should be made (updating the documentation, docstrings).

ogrisel replied:
Because it's the name of the module and I saw "pickled" used elsewhere.

I changed it because I had trouble understanding the meaning of that single letter variable when reading the code.

ogrisel replied:
Actually, the "pickled" notation was used in another project. I reverted to old notation for now.

ogrisel commented:
Alright, then I think this is good to go from my point of view.

ogrisel committed:
Typo in comment

serhiy-storchaka commented on Lib/test/pickletester.py:
```diff
@@ -2076,15 +2095,34 @@ def test_framing_many_objects(self):
+                        pickler.dump(obj)
+                        pickled = buf.getvalue()
+                    else:
+                        # Note that we cannot set the fast flag to True in
```
Then just add
```python
elif fast:
    continue
```

ogrisel committed:
Add comments for tests that require self.pickler

ogrisel commented:
Also: shall we make `frame_size_target` an additional argument to the `Pickler` class? Now that we use it to trigger flushes systematically for the C pickler this can have a significant impact on the pickling speed.

pitrou commented:
> Also: shall we make frame_size_target an additional argument to the Pickler class?

I would leave that for a separate PR. Perhaps you will want to benchmark such a change first on various kinds of IO objects.

ogrisel commented:
Sorry for the lack of reply. I will work on running some benchmark soon.

ogrisel commented:
Here are some benchmark results (for the C implementation). As expected calling write more often causes a performance regression.

On master:

```
(py37) ogrisel@ici:~/code/cpython$ ./python ~/tmp/bench_pickle.py 
# protocol 2, output: output.pkl
median time: 0.247s
# protocol 3, output: output.pkl
median time: 0.249s
# protocol 4, output: output.pkl
median time: 0.25s
# protocol 2, output: /dev/null
median time: 0.142s
# protocol 3, output: /dev/null
median time: 0.144s
# protocol 4, output: /dev/null
median time: 0.15s
last pickle size: 53.330MB
```

On this branch:

```
(py37) ogrisel@ici:~/code/cpython$ ./python ~/tmp/bench_pickle.py 
# protocol 2, output: output.pkl
median time: 0.286s
# protocol 3, output: output.pkl
median time: 0.284s
# protocol 4, output: output.pkl
median time: 0.285s
# protocol 2, output: /dev/null
median time: 0.175s
# protocol 3, output: /dev/null
median time: 0.174s
# protocol 4, output: /dev/null
median time: 0.174s
last pickle size: 53.330MB
```

ogrisel commented:
Here is the code of the benchmark:

```python
from time import time
import pickle
from statistics import median

# Nested datastructure with many small objects that should result in a
# pickle significantly larger than the target frame size:
objects = [{str(i): (bytes(i % 1000), i) for i in range(int(1e5))}]

ondisk_filename = 'output.pkl'

for filename in [ondisk_filename, '/dev/null']:
    for proto in [2, 3, 4]:
        print(f"# protocol {proto}, output: {filename}")

        timings = []
        for i in range(20):
            with open(filename, 'wb') as f:
                tic = time()
                pickle.dump(objects, f)
                toc = time()
                timings.append(toc - tic)
        print(f"median time: {median(timings):0.3}s")


with open(ondisk_filename, 'rb') as f:
    print(f"last pickle size: {len(f.read()) / 1e6:0.3f}MB")
```

ogrisel committed:
revert variable renaming

ogrisel committed:
Disable test_framed_write_sizes in OptimizedPickleTests

ogrisel committed:
Small code style fix

pitrou commented:
Interesting. What if you make the frame size larger?

ogrisel commented:
There is something fishy with the results of my benchmark: for protocol 2 and 3, there is no framing enabled, therefore the pickler still only writes once at the end of the dump. When I increase the frame size the difference is smaller but I suspect that I have other things going on my machine because I get widely different results despite taking the median time of 20 runs for each measurement.

pitrou commented:
Two suggestions:
* make the benchmark longer, to avoid effects of CPU frequency scaling
* pin the benchmark to a particular CPU core (for example using `taskset` on Linux)

pitrou commented:
I tried the benchmark on my machine and I also got a lot of variation. Sometimes the unpatched version is faster, sometimes the patched version is faster. If you want to be perfectionist, perhaps you want to try @vstinner's `perf` tool: https://perf.readthedocs.io/en/latest/

vstinner commented:
I'm disapointed that @ogrisel posted a benchmark without the standard deviation. What happened to you Olivier? :-)
http://perf.readthedocs.io/en/latest/analyze.html#minimum-vs-average

vstinner commented:
My advices to get reproductible results:
http://perf.readthedocs.io/en/latest/run_benchmark.html#how-to-get-reproductible-benchmark-results

vstinner commented on Lib/pickle.py:
```diff
@@ -202,10 +202,8 @@ def commit_frame(self, force=False):
                     write = self.file_write
-                    write(FRAME)
-                    write(pack("<Q", n))
+                    write(FRAME + pack("<Q", len(data)))
```
Can you please add a short comment explaining this change? At the first read, this change seems inefficient.

ogrisel commented:
I could have bootstrapped the median to build confidence intervals but it's not (yet?) part of the `statistics` module of the standard library :)

Anyway I used `python3 -m perf tune` to set my system in "performance" mode, I use taskset to run the bench on CPU 0 only, and disabled the GC and now I get stable benchmarks:

- On master:

```
# protocol 2, output: output.pkl
median time: 0.336s
# protocol 3, output: output.pkl
median time: 0.338s
# protocol 4, output: output.pkl
median time: 0.336s
# protocol 2, output: /dev/null
median time: 0.191s
# protocol 3, output: /dev/null
median time: 0.189s
# protocol 4, output: /dev/null
median time: 0.189s
last pickle size: 53.330MB
```

- On this branch:

```
$ taskset -c 0 ./python ~/tmp/bench_pickle.py 
# protocol 2, output: output.pkl
median time: 0.333s
# protocol 3, output: output.pkl
median time: 0.334s
# protocol 4, output: output.pkl
median time: 0.334s
# protocol 2, output: /dev/null
median time: 0.185s
# protocol 3, output: /dev/null
median time: 0.188s
# protocol 4, output: /dev/null
median time: 0.188s
last pickle size: 53.330MB
```

So there is no performance regression on large collections of small objects.

ogrisel committed:
Add comments to explain tradeoffs in Python pickle frame commit

ogrisel commented:
I think I addressed all the comments.

ogrisel committed:
Implement no-copy support for delayed writers in Python pickler

ogrisel committed:
Fix renamed test_framed_write_sizes_with_delayed_writer in test_pickletools.py

pitrou commented on Modules/_pickle.c:
```diff
@@ -2057,6 +2068,51 @@ save_float(PicklerObject *self, PyObject *obj)
+static int
+_Pickler_write_large_bytes(
+    PicklerObject *self, const char *header, Py_ssize_t header_size,
+    PyObject *payload, Py_ssize_t payload_size)
```
`payload_size` doesn't seem used below.

pitrou commented on Modules/_pickle.c:
```diff
@@ -952,6 +938,31 @@ _Pickler_FlushToFile(PicklerObject *self)
+        if(_Pickler_CommitFrame(self)) {
+            return -1;
+        }
+        if (self->write != NULL) {
```
Would be nice to add a comment here.

pitrou reviewed (changes requested):
I think this is mostly good to go now. Just two small comments.

ogrisel committed:
Add comment to explain flush after commit in _Pickler_OpcodeBoundary

ogrisel committed:
Removed unused argument payload_size

ogrisel committed:
Update the changelog to reflect the new behavior of the Python pickler

ogrisel committed:
Fix phrasing

serhiy-storchaka commented on Lib/pickle.py:
```diff
@@ -201,21 +201,46 @@ def commit_frame(self, force=False):
-                f.truncate()
+                data = f.getbuffer()
+                write = self.file_write
+                # Issue a single call to the write nethod of the underlying
```
s/nethod/method/

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -952,6 +938,38 @@ _Pickler_FlushToFile(PicklerObject *self)
+         * to limit memory usage when dumping large complex objects to
+         * a file.
+         *
+         * self-write is NULL when called via dumps.
```
`self->write`

serhiy-storchaka reviewed (approved):
LGTM! I have found just few typos.

I don't have opinion about `getbuffer`/`getvalue`. Let defer it until we encounter concrete issue.

ogrisel committed:
Fixed typos in comments
